mods/page_revisions.js

```diff
@@ -293,56 +425,72 @@ PRS.prototype.getTitleRevision = function(restbase, req) {
+                        title: title,
+                        rev: parseInt(rp.revision)
+                    },
+                    limit: 1
+                }
+            })
+            .catch(function(e) {
+                if (e.status !== 404) {
+                    throw e;
+                }
+                return self.fetchAndStoreMWRevision(restbase, req);
+            }),
+            pageData: pageDataRequest
         });
     } else if (!rp.revision) {
         if (req.headers && /no-cache/.test(req.headers['cache-control'])) {
-            revisionRequest = self.fetchAndStoreMWRevision(restbase, req)
+            revisionRequest = P.props({
+                revisionInfo: self.fetchAndStoreMWRevision(restbase, req),
+                pageData: pageDataRequest
+            })
+            .then(function(res) {
+                return self._checkPageUndeleted(restbase, req, res);
+            })
             .catch(function(e) {
                 if (e.status !== 404) {
                     throw e;
                 }
-                return getLatestTitleRevision()
-                // In case 404 is returned by MW api, the page is deleted
-                .then(function(result) {
-                    result = result.body.items[0];
-                    result.tid = uuid.now().toString();
-                    result.restrictions = result.restrictions || [];
-                    result.restrictions.push('page_deleted');
-                    return restbase.put({
-                        uri: self.tableURI(rp.domain),
-                        body: {
-                            table: self.tableName,
-                            attributes: result
-                        }
-                    })
-                    .then(function() {
-                        throw e;
-                    });
+                // Page data request never returns 404, so it's from MW API => page was deleted
+                return self._signalPageDeleted(restbase, req)
+                .then(function() {
+                    throw e;
                 });
             });
         } else {
-            revisionRequest = getLatestTitleRevision()
-            .catch(function(e) {
-                if (e.status !== 404) {
-                    throw e;
+            revisionRequest = P.props({
+                revisionInfo: getLatestTitleRevision()
+                .catch(function(e) {
+                    if (e.status !== 404) {
+                        throw e;
+                    }
+                    return self.fetchAndStoreMWRevision(restbase, req);
+                }),
+                pageData: pageDataRequest
+            })
+            .then(function(res) {
+                if (res.pageData
+                        && res.pageData.body.items
+                        && res.pageData.body.items.length) {
+                    var latestEvent = res.pageData.body.items[0];
+                    if (latestEvent.event_type === 'rename_to') {
```

Comment:
I'm wondering if this is the correct behaviour. For correctness' sake, if we say *page renamed to X* then we should ensure *X* is indeed the latest page name.

I'm also worried about the correctness of throwing a 404 at all here. I think the correct way to go about this is either *(a)* respond with a 301 and point to the latest title URI; or *(b)* supply the data about the revision and add an additional header indicating the client should use the new title. Either solution would allow us to preserve stable URIs and allow viewing historical content regardless of the point in time the URI was obtained.